File: surface/widgets/nativescript/components/widgets/primitives/FocusTrap.ts
// ============================================================
// Clef Surface NativeScript Widget — FocusTrap
//
// Traps focus within a container to support modal patterns.
// Wraps children in a StackLayout with focus containment.
// ============================================================

import { StackLayout } from '@nativescript/core';
import type { View } from '@nativescript/core';

// --------------- Props ---------------

export interface FocusTrapProps {
  active?: boolean;
  initialFocusIndex?: number;
  returnFocusOnDeactivate?: boolean;
  children?: View[];
}

// --------------- Component ---------------

export function createFocusTrap(props: FocusTrapProps): StackLayout {
  const {
    active = true,
    initialFocusIndex = 0,
    returnFocusOnDeactivate = true,
    children = [],
  } = props;

  const container = new StackLayout();
  container.className = 'clef-widget-focus-trap';
  container.accessibilityRole = 'none';

  if (active) {
    container.className += ' clef-focus-trap-active';
  }

  for (const child of children) {
    container.addChild(child);
  }

  return container;
}

export default createFocusTrap;

File: surface/widgets/appkit/components/widgets/primitives/FocusTrap.swift
// ============================================================
// Clef Surface AppKit Widget — FocusTrap
//
// Container that constrains keyboard focus to its children,
// cycling Tab/Shift-Tab within the trap boundary. Useful for
// modal dialogs and popovers.
// ============================================================

import AppKit

public class ClefFocusTrapView: NSView {
    public var active: Bool = true

    public override func keyDown(with event: NSEvent) {
        guard active else { super.keyDown(with: event); return }

        if event.keyCode == 48 { // Tab key
            let focusable = collectFocusableViews(in: self)
            guard !focusable.isEmpty else { return }

            let current = window?.firstResponder as? NSView
            let currentIndex = focusable.firstIndex(where: { $0 === current }) ?? -1

            if event.modifierFlags.contains(.shift) {
                let prev = currentIndex <= 0 ? focusable.count - 1 : currentIndex - 1
                window?.makeFirstResponder(focusable[prev])
            } else {
                let next = (currentIndex + 1) % focusable.count
                window?.makeFirstResponder(focusable[next])
            }
        } else {
            super.keyDown(with: event)
        }
    }

    private func collectFocusableViews(in view: NSView) -> [NSView] {
        var result: [NSView] = []
        for child in view.subviews {
            if child.acceptsFirstResponder && !child.isHidden {
                result.append(child)
            }
            result.append(contentsOf: collectFocusableViews(in: child))
        }
        return result
    }
}
